core/mempool/src/lib.rs

```diff
@@ -0,0 +1,112 @@
+use std::collections::HashSet;
+use std::sync::Arc;
+use parking_lot::RwLock;
+use tokio::sync::mpsc::{Sender, Receiver};
+use futures::{Future, Stream};
+use log::error;
+use primitives::transaction::SignedTransaction;
+use primitives::chain::{ReceiptBlock, ChainPayload, SignedShardBlock};
+
+/// mempool that stores transactions and receipts for a chain
+#[derive(Default)]
+struct Pool {
+    transactions: HashSet<SignedTransaction>,
+    receipts: HashSet<ReceiptBlock>,
+}
+
+impl Pool {
+    fn add_transaction(&mut self, transaction: SignedTransaction) {
+        self.transactions.insert(transaction);
+    }
+
+    fn add_receipt(&mut self, receipt: ReceiptBlock) {
+        self.receipts.insert(receipt);
+    }
+
+    #[allow(unused)]
+    fn produce_payload(&mut self) -> ChainPayload {
+        let transactions: Vec<_> = self.transactions.drain().collect();
+        let receipts: Vec<_> = self.receipts.drain().collect();
+        ChainPayload { transactions, receipts }
+    }
+
+    fn import_block(&mut self, block: &SignedShardBlock) {
```

Comment:
Why block importing removes transactions?

Reply:
Because imported block already contains it, which means it shouldn't be suggested for next block.
Import block here is only ran if finalzied.